expression/util.go

```diff
@@ -240,3 +241,70 @@ func ConvertCol2CorCol(cond Expression, corCols []*CorrelatedColumn, outerSchema
 	}
 	return cond
 }
+
+// ConvertConstantType convert constant value to appropriate type
+// See https://dev.mysql.com/doc/refman/5.7/en/type-conversion.html for detail
+func ConvertConstantType(expr Expression) (Expression, error) {
+	scalarFunc, ok := expr.(*ScalarFunction)
+	if !ok {
+		return expr, nil
+	}
+
+	var err error
+	args := scalarFunc.GetArgs()
+
+	switch scalarFunc.FuncName.L {
+	case ast.EQ, ast.GT, ast.LE, ast.GE, ast.LT:
+		for i, arg := range args {
+			if c, ok := arg.(*Constant); ok { // check if there is a constant
+				tp := args[(i+1)%2].GetType() // only two args here
+				err = convertConstantTo(tp, c)
+			} else {
+				args[i], err = ConvertConstantType(arg)
+			}
+			if err != nil {
+				break
+			}
+		}
+	default:
+		for i, arg := range args {
+			if args[i], err = ConvertConstantType(arg); err != nil {
+				break
+			}
+		}
+	}
+
+	return scalarFunc, errors.Trace(err)
+}
+
+func convertConstantTo(tp *types.FieldType, constant *Constant) error {
+	if tp == nil {
+		return nil
+	}
+	switch tp.Tp { // check if need to convert type based on tp
+	case mysql.TypeDatetime, mysql.TypeTimestamp:
+		return convertConstantToTime(tp, constant)
+		// add miss case here in the future
+	default:
+		return nil
+	}
+}
+
+func convertConstantToTime(tp *types.FieldType, constant *Constant) error {
+	var (
+		t   types.Time
+		err error
+	)
+	switch constant.Value.Kind() {
+	case types.KindInt64, types.KindUint64:
```

Comment:
Same as MySQL, return all rows in this case.

But I found a bug:
 `select cast(20021020.1111 as datetime)`
result is `[HY000][1105] cannot convert &{8 0 0 0 0 [] 0xc420bfb3e0} to type datetime`

I think we need make the type conversion module more robust.